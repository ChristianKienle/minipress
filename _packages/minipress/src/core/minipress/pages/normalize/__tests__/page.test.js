// @ts-check

const { normalizePage } = require('./../')
const Transformers = require('./../../../transformers')
const ContentComponents = require('./../../../content-components')
const TempDir = require('./../../../temp-dir')
const fs = require('fs-extra')
const OS = require('os')
const nanoid = require('nanoid')
const { join } = require('path')

const createTempDir = () => {
  const path = join(OS.tmpdir(), nanoid())
  fs.mkdirSync(path)
  let cleaned = false
  const cleanup = () => {
    if (cleaned === false) {
      fs.removeSync(path)
    }
    cleaned = true
  }
  const tempDir = new TempDir({ path })
  return {
    tempDir,
    cleanup
  }
}

const MarkdownTransformer = require('@minipress/plugin-format-markdown/src/transformer')

describe('normalize page', () => {
  let { tempDir, cleanup } = createTempDir()
  beforeEach(() => {
    cleanup()
    const dir = createTempDir()
    tempDir = dir.tempDir
    cleanup = dir.cleanup
  })

  afterEach(() => {
    cleanup()
  })

  it('virtual page causes content component to be created', async () => {
    const page = {
      key: 'virtual',
      content: '# hi',
      contentType: 'md',
    }

    const transformers = new Transformers()
    transformers.set('md', MarkdownTransformer())
    const contentComponents = new ContentComponents()
    const _page = await normalizePage(page, {
      transformers, contentComponents, tempDir
    })

    expect(contentComponents.get(_page.key)).toBeDefined()
    expect(_page).toMatchObject(_page)
  })

  it('layout/key/path is read from frontmatter', async () => {
    const content = '---\nlayout: custom\npath: /path\nkey: mykey\n---\n\n# hi'
    const contentType = 'md'
    const page = {
      content,
      contentType
    }

    const transformers = new Transformers()
    transformers.set('md', MarkdownTransformer())
    const contentComponents = new ContentComponents()
    const _page = await normalizePage(page, {
      transformers, contentComponents, tempDir
    })

    // expect(contentComponents.get(page.key)).toMatchObject(page)
    expect(_page).toMatchObject({
      content: '<h1 id=\"hi\"><router-link class=\"header-anchor\" to=\"#hi\" aria-hidden=\"true\">#</router-link> hi</h1>\n',
      contentType,
      key: 'mykey',
      path: '/path',
      layout: 'custom'
    })
  })
})