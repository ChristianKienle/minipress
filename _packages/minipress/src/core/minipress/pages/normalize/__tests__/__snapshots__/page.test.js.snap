// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "content": "# hi",
  "contentType": "md",
  "key": "virtual",
}
`;

exports[`normalize page virtual page 1`] = `
Object {
  "content": "# hi",
  "contentType": "md",
  "key": "virtual",
}
`;
